### ISSUE TITLE
============================
# MOTOR DE AUDITORÍA DE RUPTURAS ESTRUCTURALES — MODO EN VIVO
# PCA causal + Z-score causal + Entropía Shannon + Entropía Hugo
#===

### ISSUE THREAD

**Issue by edduchzsant21-prog** [edduchzsant21-prog/detector-de-rupturas#3]
# ================================ ============================================================
# !pip install yfinance scipy scikit-learn matplotlib -q

import os
import time
import datetime as dt

import numpy as np
import pandas as pd
import yfinance as yf
import matplotlib.pyplot as plt

from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from scipy.signal import hilbert

import warnings
warnings.filterwarnings("ignore")

# ------------------------------------------------------------
# UTILIDADES DE DESCARGA
# ------------------------------------------------------------
def descargar_precios(ticker, period="5y", interval="1d"):
    """
    Descarga precios y devuelve SIEMPRE una Serie 1-D con índice datetime.
    auto_adjust=False conserva la columna 'Adj Close'.
    """
    datos = yf.download(
        ticker,
        period=period,
        interval=interval,
        auto_adjust=False,
        progress=False
    )

    if datos.empty:
        raise ValueError(f"Sin datos para {ticker}.")

    columna = "Adj Close" if "Adj Close" in datos.columns else "Close"
    precios = datos[columna]

    if isinstance(precios, pd.DataFrame):
        precios = precios.iloc[:, 0]

    precios.index = pd.to_datetime(precios.index)
    precios = precios.sort_index().dropna().astype(float)
    precios.name = ticker

    return precios

class MotorAuditoriaRupturas:
    """
    Detección y auditoría de cambios de régimen en series financieras.

    Modo histórico: ejecutar()
    Modo en vivo  : preparar_estado() -> procesar_tick() -> ejecutar_en_vivo()
    """

    def __init__(
        self,
        horizontes=None,
        ventana_zscore=252,
        ventana_entrenamiento=252,
        ventana_minima=126,
        umbral=2.58
    ):
        self.horizontes = horizontes or [5, 10, 21, 63]
        self.ventana_zscore = ventana_zscore
        self.ventana_entrenamiento = ventana_entrenamiento
        self.ventana_minima = ventana_minima
        self.umbral = umbral

        # Estado en vivo
        self.precios = pd.Series(dtype=float)
        self.historial_pc1 = pd.Series(dtype=float)
        self.ultima_barra = None
        self.eventos_vivo = []

    # --------------------------------------------------------
    # 1. CARACTERÍSTICAS
    # --------------------------------------------------------
    def generar_caracteristicas(self, precios):
        precios = precios.dropna().astype(float)
        retornos = np.log(precios / precios.shift(1))
        df = pd.DataFrame(index=precios.index)

        for h in self.horizontes:
            df[f"ret_{h}"] = retornos.rolling(h).sum()
            df[f"vol_{h}"] = retornos.rolling(h).std()

        return df.dropna()

    # --------------------------------------------------------
    # 2. PCA CAUSAL Y Z-SCORE CAUSAL (histórico)
    # --------------------------------------------------------
    def detectar_rupturas(self, df_features, precios):
        registros = []
        n_obs = len(df_features)

        for i in range(self.ventana_minima, n_obs):
            inicio = max(0, i - self.ventana_entrenamiento)
            datos_pasados = df_features.iloc[inicio:i]
            dato_actual = df_features.iloc[[i]]

            if len(datos_pasados) < self.ventana_minima:
                continue

            scaler = StandardScaler()
            pca = PCA(n_components=1)

            X_pasado = scaler.fit_transform(datos_pasados)
            X_actual = scaler.transform(dato_actual)

            pca.fit(X_pasado)
            pc1 = pca.transform(X_actual)[0, 0]

            fecha = df_features.index[i]
            registros.append({
                "Fecha": fecha,
                "PC1": float(pc1),
                "Precio": float(precios.loc[fecha])
            })

        if not registros:
            return pd.DataFrame()

        df_res = pd.DataFrame(registros).set_index("Fecha")

        media = (df_res["PC1"]
                 .rolling(self.ventana_zscore, min_periods=self.ventana_minima)
                 .mean().shift(1))
        desv = (df_res["PC1"]
                .rolling(self.ventana_zscore, min_periods=self.ventana_minima)
                .std().shift(1))

        df_res["Z_score"] = (df_res["PC1"] - media) / desv
        return df_res.dropna()

    # --------------------------------------------------------
    # 3. ENTROPÍA SHANNON
    # --------------------------------------------------------
    @staticmethod
    def calcular_entropia_shannon(serie, n_bins=6):
        x = pd.Series(serie).dropna().values
        if len(x) < 5:
            return np.nan

        bins = min(n_bins, max(3, int(np.sqrt(len(x)))))
        hist, _ = np.histogram(x, bins=bins)

        if hist.sum() == 0:
            return np.nan

        p = hist / hist.sum()
        p = p[p > 0]
        return float(-np.sum(p * np.log2(p)))

    # --------------------------------------------------------
    # 4. ENTROPÍA HUGO (fase de Hilbert)
    # --------------------------------------------------------
    @staticmethod
    def calcular_entropia_hugo(serie, n_bins=8):
        x = pd.Series(serie).dropna().values
        if len(x) < 8:
            return np.nan

        xc = x - np.mean(x)
        if np.std(xc) == 0:
            return 0.0

        theta = np.mod(np.angle(hilbert(xc)), 2 * np.pi)
        bins = min(n_bins, max(4, int(np.sqrt(len(x)))))
        hist, _ = np.histogram(theta, bins=bins, range=(0, 2 * np.pi))

        if hist.sum() == 0:
            return np.nan

        p = hist / hist.sum()
        p = p[p > 0]
        return float(-np.sum(p * np.log2(p)) / np.log2(bins))

    # --------------------------------------------------------
    # 5. AGRUPAR EVENTOS
    # --------------------------------------------------------
    @staticmethod
    def filtrar_eventos(eventos, dias_exclusion=10):
        if eventos.empty:
            return eventos.copy()

        ev = eventos.copy()
        ev["abs_z"] = ev["Z_score"].abs()
        ev = ev.sort_values("abs_z", ascending=False)

        seleccionadas = []
        for fecha in ev.index:
            cercano = any(
                abs((fecha - f0).days) <= dias_exclusion
                for f0 in seleccionadas
            )
            if not cercano:
                seleccionadas.append(fecha)

        return ev.loc[seleccionadas].sort_index().drop(columns="abs_z")

    # --------------------------------------------------------
    # 6. AUDITORÍA PRE/POST
    # --------------------------------------------------------
    def auditar_entorno_eventos(
        self, precios, df_resultados,
        umbral=None, ventana_dias=21, dias_exclusion=10
    ):
        umbral = self.umbral if umbral is None else umbral
        if df_resultados.empty:
            return pd.DataFrame()

        precios = precios.dropna().astype(float)
        retornos = np.log(precios / precios.shift(1)).dropna()

        brutos = df_resultados[df_resultados["Z_score"].abs() > umbral].copy()
        eventos = self.filtrar_eventos(brutos, dias_exclusion)

        filas = []
        for fecha, fila in eventos.iterrows():
            if fecha not in retornos.index:
                continue

            pos = retornos.index.get_loc(fecha)
            pre = retornos.iloc[pos - ventana_dias:pos]
            post = retornos.iloc[pos + 1:pos + 1 + ventana_dias]

            if len(pre) < ventana_dias or len(post) < ventana_dias:
                continue

            s_pre = self.calcular_entropia_shannon(pre)
            s_post = self.calcular_entropia_shannon(post)
            h_pre = self.calcular_entropia_hugo(pre)
            h_post = self.calcular_entropia_hugo(post)

            filas.append({
                "Fecha_evento": fecha,
                "Precio_evento": float(fila["Precio"]),
                "PC1": float(fila["PC1"]),
                "Z_score": float(fila["Z_score"]),
                "Direccion": "Ruptura positiva" if fila["Z_score"] > 0 else "Ruptura negativa",
                "Shannon_pre": s_pre,
                "Shannon_post": s_post,
                "Delta_Shannon": s_post - s_pre,
                "Hugo_pre": h_pre,
                "Hugo_post": h_post,
                "Delta_Hugo": h_post - h_pre,
                "Volatilidad_pre": float(pre.std()),
                "Volatilidad_post": float(post.std()),
                "Delta_volatilidad": float(post.std() - pre.std()),
                "Retorno_pre": float(pre.sum()),
                "Retorno_post": float(post.sum())
            })

        if not filas:
            return pd.DataFrame()

        return pd.DataFrame(filas).set_index("Fecha_evento")

    # --------------------------------------------------------
    # 7. MONTE CARLO
    # --------------------------------------------------------
    @staticmethod
    def prueba_montecarlo(precios, auditoria, ventana_dias=21,
                          n_simulaciones=500, semilla=42):
        if auditoria.empty:
            return pd.DataFrame()

        precios = precios.dropna().astype(float)
        retornos = np.log(precios / precios.shift(1)).dropna()
        rng = np.random.default_rng(semilla)

        obs = {
            "shannon": auditoria["Delta_Shannon"].mean(),
            "hugo": auditoria["Delta_Hugo"].mean(),
            "vol": auditoria["Delta_volatilidad"].mean()
        }

        validas = retornos.index[ventana_dias:len(retornos) - ventana_dias - 1]
        n_ev = min(len(auditoria), len(validas))

        sim_s, sim_h, sim_v = [], [], []

        for _ in range(n_simulaciones):
            fechas = rng.choice(validas, size=n_ev, replace=False)
            ds, dh, dv = [], [], []

            for f in fechas:
                pos = retornos.index.get_loc(f)
                pre = retornos.iloc[pos - ventana_dias:pos]
                post = retornos.iloc[pos + 1:pos + 1 + ventana_dias]

                ds.append(MotorAuditoriaRupturas.calcular_entropia_shannon(post)
                          - MotorAuditoriaRupturas.calcular_entropia_shannon(pre))
                dh.append(MotorAuditoriaRupturas.calcular_entropia_hugo(post)
                          - MotorAuditoriaRupturas.calcular_entropia_hugo(pre))
                dv.append(post.std() - pre.std())

            sim_s.append(np.nanmean(ds))
            sim_h.append(np.nanmean(dh))
            sim_v.append(np.nanmean(dv))

        sim_s, sim_h, sim_v = map(np.array, (sim_s, sim_h, sim_v))

        def pval(sim, valor):
            return (np.sum(np.abs(sim) >= abs(valor)) + 1) / (n_simulaciones + 1)

        return pd.DataFrame({
            "Metrica": ["Delta_Shannon", "Delta_Hugo", "Delta_volatilidad"],
            "Valor_observado": [obs["shannon"], obs["hugo"], obs["vol"]],
            "Media_nula": [np.nanmean(sim_s), np.nanmean(sim_h), np.nanmean(sim_v)],
            "Std_nula": [np.nanstd(sim_s), np.nanstd(sim_h), np.nanstd(sim_v)],
            "P_value_empirico": [pval(sim_s, obs["shannon"]),
                                 pval(sim_h, obs["hugo"]),
                                 pval(sim_v, obs["vol"])]
        })

    # --------------------------------------------------------
    # 8. MODO HISTÓRICO
    # --------------------------------------------------------
    def ejecutar(self, precios, umbral=None, ventana_dias=21, dias_exclusion=10):
        umbral = self.umbral if umbral is None else umbral
        features = self.generar_caracteristicas(precios)
        resultados = self.detectar_rupturas(features, precios)
        auditoria = self.auditar_entorno_eventos(
            precios, resultados, umbral, ventana_dias, dias_exclusion
        )
        return features, resultados, auditoria

    # ========================================================
    # 9. MODO EN VIVO
    # ========================================================
    def preparar_estado(self, precios_historicos):
        """
        Calienta el motor con el histórico: guarda precios y la serie PC1
        necesaria para el Z-score causal del próximo tick.
        """
        self.precios = precios_historicos.copy()
        features = self.generar_caracteristicas(self.precios)
        resultados = self.detectar_rupturas(features, self.precios)

        if resultados.empty:
            raise ValueError("Histórico insuficiente para calentar el motor.")

        self.historial_pc1 = resultados["PC1"].copy()
        self.ultima_barra = self.precios.index[-1]

        print(f"Motor calentado con {len(self.precios)} barras "
              f"(hasta {self.ultima_barra.date()}).")
        return resultados

    def _pc1_ultimo(self, features):
        """PCA causal aplicado solo a la última observación."""
        i = len(features) - 1
        inicio = max(0, i - self.ventana_entrenamiento)
        pasados = features.iloc[inicio:i]

        if len(pasados) < self.ventana_minima:
            return np.nan

        scaler = StandardScaler()
        X_pasado = scaler.fit_transform(pasados)
        X_actual = scaler.transform(features.iloc[[i]])

        pca = PCA(n_components=1)
        pca.fit(X_pasado)
        return float(pca.transform(X_actual)[0, 0])

    def procesar_tick(self, fecha, precio):
        """
        Incorpora una nueva observación y devuelve el diagnóstico.
        Devuelve None si la barra ya fue procesada.
        """
        fecha = pd.to_datetime(fecha)

        if self.ultima_barra is not None and fecha <= self.ultima_barra:
            return None  # barra repetida

        self.precios.loc[fecha] = float(precio)
        self.precios = self.precios.sort_index()
        self.ultima_barra = fecha

        features = self.generar_caracteristicas(self.precios)
        if features.empty:
            return None

        pc1 = self._pc1_ultimo(features)
        if np.isnan(pc1):
            return None

        # Z-score causal: estadísticos SIN incluir el punto actual
        ventana = self.historial_pc1.tail(self.ventana_zscore)
        if len(ventana) < self.ventana_minima:
            z = np.nan
        else:
            mu, sd = ventana.mean(), ventana.std()
            z = (pc1 - mu) / sd if sd > 0 else np.nan

        self.historial_pc1.loc[fecha] = pc1

        retornos = np.log(self.precios / self.precios.shift(1)).dropna()
        ultimos = retornos.tail(21)

        estado = {
            "Fecha": fecha,
            "Precio": float(precio),
            "PC1": pc1,
            "Z_score": float(z) if not np.isnan(z) else np.nan,
            "Shannon_21d": self.calcular_entropia_shannon(ultimos),
            "Hugo_21d": self.calcular_entropia_hugo(ultimos),
            "Volatilidad_21d": float(ultimos.std()),
            "Alerta": bool(not np.isnan(z) and abs(z) > self.umbral),
            "Direccion": ("Ruptura positiva" if z > 0 else "Ruptura negativa")
                         if (not np.isnan(z) and abs(z) > self.umbral) else "Normal"
        }

        if estado["Alerta"]:
            self.eventos_vivo.append(estado)

        return estado

    def ejecutar_en_vivo(
        self,
        ticker,
        intervalo_datos="1d",
        periodo_calentamiento="5y",
        segundos_espera=60,
        max_iteraciones=None,
        archivo_log="monitor_rupturas.csv",
        verbose=True
    ):
        """
        Bucle de monitoreo continuo.

        intervalo_datos : '1m', '5m', '15m', '1h', '1d'
        segundos_espera : frecuencia de sondeo
        max_iteraciones : None = indefinido; útil para pruebas
        """
        # Calentamiento
        if intervalo_datos == "1d":
            hist = descargar_precios(ticker, periodo_calentamiento, "1d")
        else:
            hist = descargar_precios(ticker, "60d", intervalo_datos)

        self.preparar_estado(hist)

        iteracion = 0
        print(f"\nMonitoreo en vivo: {ticker} | intervalo {intervalo_datos} "
              f"| sondeo cada {segundos_espera}s\nCtrl+C para detener.\n")

        try:
            while max_iteraciones is None or iteracion < max_iteraciones:
                iteracion += 1
                try:
                    periodo = "5d" if intervalo_datos != "1d" else "1mo"
                    reciente = descargar_precios(ticker, periodo, intervalo_datos)
                    nuevas = reciente[reciente.index > self.ultima_barra]

                    for fecha, precio in nuevas.items():
                        estado = self.procesar_tick(fecha, precio)
                        if estado is None:
                            continue

                        linea = (
                            f"[{estado['Fecha']}] "
                            f"P={estado['Precio']:.2f} | "
                            f"PC1={estado['PC1']:+.3f} | "
                            f"Z={estado['Z_score']:+.2f} | "
                            f"S={estado['Shannon_21d']:.3f} | "
                            f"H={estado['Hugo_21d']:.3f} | "
                            f"{estado['Direccion']}"
                        )
                        if estado["Alerta"]:
                            print("*** ALERTA *** " + linea)
                        elif verbose:
                            print(linea)

                        pd.DataFrame([estado]).to_csv(
                            archivo_log,
                            mode="a",
                            header=not os.path.exists(archivo_log),
                            index=False
                        )

                    if verbose and nuevas.empty:
                        ahora = dt.datetime.now().strftime("%H:%M:%S")
                        print(f"[{ahora}] sin barras nuevas.")

                except Exception as e:
                    print(f"Error en el sondeo {iteracion}: {e}")

                time.sleep(segundos_espera)

        except KeyboardInterrupt:
            print("\nMonitoreo detenido por el usuario.")

        return pd.DataFrame(self.eventos_vivo)

    # --------------------------------------------------------
    # 10. GRÁFICAS
    # --------------------------------------------------------
    @staticmethod
    def graficar_resultados(df_resultados, auditoria, ticker="Activo", umbral=2.58):
        if df_resultados.empty:
            print("No hay suficientes datos para graficar.")
            return

        fig, axes = plt.subplots(
            3, 1, figsize=(16, 12), sharex=True,
            gridspec_kw={"height_ratios": [2, 1, 1]}
        )

        axes[0].plot(df_resultados.index, df_resultados["Precio"],
                     color="navy", linewidth=1.1, label="Precio")
        if auditoria is not None and not auditoria.empty:
            axes[0].scatter(auditoria.index, auditoria["Precio_evento"],
                            color="red", marker="v", s=90,
                            label="Ruptura", zorder=5)
        axes[0].set_title(f"{ticker} — Auditoría de Rupturas Estructurales")
        axes[0].set_ylabel("Precio"); axes[0].grid(alpha=.25); axes[0].legend()

        axes[1].plot(df_resultados.index, df_resultados["PC1"],
                     color="darkorange", linewidth=1, label="PC1 causal")
        axes[1].axhline(0, color="black", linewidth=.8)
        axes[1].set_ylabel("PC1"); axes[1].grid(alpha=.25); axes[1].legend()

        axes[2].plot(df_resultados.index, df_resultados["Z_score"],
                     color="purple", linewidth=1, label="Z-score causal")
        axes[2].axhline(umbral, color="red", ls="--", lw=1)
        axes[2].axhline(-umbral, color="red", ls="--", lw=1)
        axes[2].set_ylabel("Z-score"); axes[2].set_xlabel("Fecha")
        axes[2].grid(alpha=.25); axes[2].legend()

        plt.tight_layout()
        plt.show()
